File: src/main/java/com/zepben/cimbend/network/tracing/PhaseStep.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.cimbend.network.tracing;

import com.google.common.collect.ImmutableSet;
import com.zepben.annotations.EverythingIsNonnullByDefault;
import com.zepben.cimbend.cim.iec61970.base.core.ConductingEquipment;
import com.zepben.cimbend.cim.iec61970.base.core.PhaseCode;
import com.zepben.cimbend.cim.iec61970.base.wires.SinglePhaseKind;

import javax.annotation.Nullable;
import java.util.Collection;
import java.util.Objects;
import java.util.Set;

/**
 * Class that records which cores were traced to get to a given conducting equipment during a trace.
 * Allows a trace to continue only on the cores used to get to the current step in the trace.
 */
@SuppressWarnings("WeakerAccess")
@EverythingIsNonnullByDefault
public class PhaseStep {

    private final ConductingEquipment conductingEquipment;
    private final Set<SinglePhaseKind> phases;
    @Nullable private final ConductingEquipment previous;

    public static PhaseStep startAt(ConductingEquipment conductingEquipment, Collection<SinglePhaseKind> phases) {
        return new PhaseStep(conductingEquipment, phases, null);
    }

    public static PhaseStep startAt(ConductingEquipment conductingEquipment, PhaseCode phaseCode) {
        return startAt(conductingEquipment, phaseCode.singlePhases());
    }

    public static PhaseStep continueAt(ConductingEquipment conductingEquipment, Collection<SinglePhaseKind> phases, ConductingEquipment previous) {
        return new PhaseStep(conductingEquipment, phases, previous);
    }

    public static PhaseStep continueAt(ConductingEquipment conductingEquipment, PhaseCode phaseCode, ConductingEquipment previous) {
        return continueAt(conductingEquipment, phaseCode.singlePhases(), previous);
    }

    public ConductingEquipment conductingEquipment() {
        return conductingEquipment;
    }

    public Set<SinglePhaseKind> phases() {
        return phases;
    }

    @Nullable
    public ConductingEquipment previous() {
        return previous;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof PhaseStep)) return false;
        PhaseStep that = (PhaseStep) o;
        return Objects.equals(conductingEquipment, that.conductingEquipment) &&
            Objects.equals(phases, that.phases);
    }

    @Override
    public int hashCode() {
        return Objects.hash(conductingEquipment, phases);
    }

    @Override
    public String toString() {
        return "PhaseStep{" +
            "current = " + conductingEquipment.getMRID() +
            ", phases = " + phases +
            ", previous = " + previous +
            '}';
    }

    private PhaseStep(ConductingEquipment conductingEquipment, Collection<SinglePhaseKind> phases, @Nullable ConductingEquipment previous) {
        this.conductingEquipment = conductingEquipment;
        this.phases = ImmutableSet.copyOf(phases);
        this.previous = previous;
    }

}

File: src/main/kotlin/com/zepben/evolve/services/network/tracing/phases/PhaseStep.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.evolve.services.network.tracing.phases

import com.zepben.evolve.cim.iec61970.base.core.ConductingEquipment
import com.zepben.evolve.cim.iec61970.base.core.PhaseCode
import com.zepben.evolve.cim.iec61970.base.wires.SinglePhaseKind
import java.util.*

/**
 * Class that records which cores were traced to get to a given conducting equipment during a trace.
 * Allows a trace to continue only on the cores used to get to the current step in the trace.
 */
class PhaseStep private constructor(
    val conductingEquipment: ConductingEquipment,
    phases: Collection<SinglePhaseKind>,
    val previous: ConductingEquipment?
) {

    val phases: Set<SinglePhaseKind> = phases.toSet()

    override fun equals(other: Any?): Boolean = when {
        this === other -> true
        other !is PhaseStep -> false
        else -> conductingEquipment == other.conductingEquipment &&
            phases == other.phases
    }

    override fun hashCode(): Int = Objects.hash(conductingEquipment, phases)

    override fun toString(): String =
        "PhaseStep{" +
            "current = " + conductingEquipment.mRID +
            ", phases = " + phases +
            ", previous = " + previous +
            '}'

    companion object {

        @JvmStatic
        fun startAt(conductingEquipment: ConductingEquipment, phases: Collection<SinglePhaseKind>): PhaseStep =
            PhaseStep(conductingEquipment, phases, null)

        @JvmStatic
        fun startAt(conductingEquipment: ConductingEquipment, phaseCode: PhaseCode): PhaseStep =
            startAt(conductingEquipment, phaseCode.singlePhases)

        @JvmStatic
        fun continueAt(conductingEquipment: ConductingEquipment, phases: Collection<SinglePhaseKind>, previous: ConductingEquipment?): PhaseStep =
            PhaseStep(conductingEquipment, phases, previous)

        @JvmStatic
        fun continueAt(conductingEquipment: ConductingEquipment, phaseCode: PhaseCode, previous: ConductingEquipment?): PhaseStep =
            continueAt(conductingEquipment, phaseCode.singlePhases, previous)

    }

}
